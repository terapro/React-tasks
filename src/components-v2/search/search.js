import React from 'react';
import {Title} from 'src/components-v2/common/title/title';
import {Input} from 'src/components-v2/common/input/input';
import {TypeSection} from 'src/components-v2/search/type-section/type-section';
import {ButtonSection} from 'src/components-v2/search/button-section/button-section';
import {parameters} from 'src/data/constants.js';

import history from 'src/store/browserHistory';

export class Search extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      activeSearchType: parameters.searchTypes[0].value,
      currentInputValue: '',
      sortByParam: parameters.sortParameters[0].value // TODO: load this value from the store
    }

  }
  changeSearchType = (newType) => {
    this.setState({activeSearchType: newType});
  }
  cloneSearchInput = (el) => {
    this.setState({currentInputValue: el.target.value});
  };
  startSearching = () => {
  //  this.props.onSearch(this.state.currentInputValue, this.state.searchTypes.active);
    const phrase = this.state.currentInputValue;
    const type = this.state.activeSearchType;
    const sortBy = this.state.sortByParam;
    const offset = 0;
    const query = '/search?phrase=' + phrase + '&searchBy=' + type + '&sortBy=' + sortBy + '&offset=' + offset;

    console.log(this.props.location);
    this.setState({currentInputValue: ''});
  //  window.location('/search?info=' + phrase');
  //  location.search = 'info=' + phrase;
   history.replace(query);
   history.go(0);

  };
  enterKeyPressedOnInput = (e) => {
    if (e.key === 'Enter') {
      this.startSearching();
    }
  };

  render() {

    return (
      <div className='search'>
        <Title content='Find your movie' ttlWhite ttlUC />
        <Input onStartTyping={this.cloneSearchInput}
                 value={this.state.currentInputValue}
                 onKeyEnterPressed={this.enterKeyPressedOnInput}/>
        <div className='search-components'>
          <TypeSection activeSearchType={this.state.activeSearchType} searchTypeClick={this.changeSearchType}/>
          <ButtonSection searchButtonClick={this.startSearching}/>
        </div>
      </div>
    );
  }
};
