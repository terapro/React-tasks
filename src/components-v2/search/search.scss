div.search {
  margin: $elementPadding 0;
}

div.search-components {
  display: flex;
  margin-top: 10px;
  justify-content: space-between;
  align-items: center;
}
