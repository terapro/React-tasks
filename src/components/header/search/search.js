import React, {Component} from 'react';

import {Input} from '../../search/input/input.js';
import {TypeSection} from '../../search/type-section/type-section.js';
import {ButtonSection} from '../../search/button-section/button-section.js';

class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchTypesActivityList: {
        'title': true,
        'genre': false
      }
    };
  }
  render() {
    return (
      <div className={'header-search'} id={'search-form'}>
        <Input placeholder={'Start typing...'}/>
        <div className={'search-components'}>
          <TypeSection searchTypes = {this.state.searchTypesActivityList} name = {'searchby'}/>
          <ButtonSection parentFormId={'search-form'}/>
        </div>
      </div>
    )
  }
}

export {Search};