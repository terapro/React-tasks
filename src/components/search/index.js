import React, {Component} from 'react';

import PropTypes from 'prop-types';

import {Input} from '../common/input/input.js';
import {Title} from '../common/title/title.js';
import {TypeSection} from '../search/type-section/type-section.js';
import {ButtonSection} from '../search/button-section/button-section.js';

class Search extends Component {
  constructor(props) {
    super(props);

    this.state = {
      searchTypes: {
        list: ['title', 'genre'],
        active: 'title'
      },
      searchPhrase: ''

    };

    this.searchButtonClick = () => {// CallBack for SearchButton click
      this.props.startSearch({searchPhrase: this.state.searchPhrase, searchType: this.state.searchTypes.active});
      this.setState({searchPhrase: ''});
    };
    this.enterKeyPressedOnInput = (e) => {
      if (e.key === 'Enter') {
        this.searchButtonClick();
      }
    };
    this.changeSearchType = (el) => {
      this.setState({searchTypes: {
          list: this.state.searchTypes.list, // living as it is
          active: el.target.id // Toggle to the active type
        }});
    };
    this.cloneSearchInput = (el) => {
      this.setState({searchPhrase: el.target.value});
    }
  }

  render() {
    console.log('Rendered!');
    return (
      <div className={'header-search'} id={'search-form'}>
        <Title  content='Find your movie' ttlWhite={true} ttlUC={true}/>
        <Input searchInputCallback = {this.cloneSearchInput} value = {this.state.searchPhrase} onKeyPressedFunc={this.enterKeyPressedOnInput} />
        <div className={'search-components'}>
          <TypeSection searchTypes = {this.state.searchTypes} name = {'searchby'}
                       searchTypeCallback = {this.changeSearchType}/>
          <ButtonSection parentFormId={'search-form'} searchButtonCallBack = {this.searchButtonClick}/>
        </div>
      </div>
    )
  }
}
Search.propTypes = {
  startSearch: PropTypes.func
};

export {Search};