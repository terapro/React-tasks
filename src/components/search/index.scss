.header-search {
  margin-top: 100px;
}
.search-components {
  display: flex;
  margin-top: 10px;
  justify-content: space-between;
  align-items: center;
}

