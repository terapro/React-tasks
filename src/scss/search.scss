$searchInputPaddingHorizontal: 20px;


.search-input {
  width: $headerContainerWidth - 2 * $searchInputPaddingHorizontal;
  background-color: #000;
  border: 0px;
  border-bottom: 2px solid $color_2;
  height: 45px;
  color: #EAEAEA;
  padding: 0 $searchInputPaddingHorizontal;
  font-size: 20px;
  line-height: 45px;
}

.search-type-section {


}

.search-button-section {


}

.search-radio-input {
display: none;
}

.search-radio-label {
  display: inline-block;
  line-height: 30px;
  background: $color_3;
  cursor: pointer;
  border-radius: 5px;
  font-size: 14px;
  text-transform: uppercase;
  letter-spacing: 1px;
  padding: 0 20px;
  margin: 0 0 0 15px;
  font-weight: normal;

}
.search-radio-input:checked + label {
  background: $color_2;

}