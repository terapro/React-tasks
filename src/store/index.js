import { createStore, combineReducers, applyMiddleware } from 'redux';
import {films, mode, search} from 'src/store/reducers';
import stateData from 'src/data/initialState.json'
import C from 'src/constants';
import {resultsToStore, recommendedToStore} from 'src/actions'; // TODO; take another action


/**
 * For API requist to the server
 * @param {string} phrase
 * @param {string} searchBy
 * @param {number} limit
 * @returns {Promise<any>}
 * @constructor
 */
const RequestToServer = (phrase, searchBy, limit=21) =>

  new Promise((resolves, rejects) => {
    console.log(phrase);
    console.log(searchBy);
    const api = `http://react-cdp-api.herokuapp.com/movies?search=${phrase}&searchBy=${searchBy}&limit=${limit}`;
    console.log(api);
    const request = new XMLHttpRequest();
    request.open('GET', api);
    request.onload = () =>
      (request.status === 200) ?
        resolves(JSON.parse(request.response)) :
        reject(Error(request.statusText));
    request.onerror = (err) => rejects(err);
    request.send();
  });


const logger = store => next => action => {

  console.log('Before ', action.type);

  // Starting API request for Search mode
  if(action.type == C.START_SEARCH) {// Request to the server
    RequestToServer(action.payload.searchPhrase, action.payload.searchType).then(
      info => store.dispatch(resultsToStore(info.data)),
      err => console.error(
        new Error("cannot load films from server"))
    );
  }
  // Starting API request for Film mode
  if(action.type == C.OPEN_FILM) {// Request to the server
    RequestToServer(action.payload.filmGenre, action.payload.searchType).then(
      info => store.dispatch(recommendedToStore(info.data)),
      err => console.error(
        new Error("cannot load films from server"))
    );
  }

  const returnVal = next(action);
  console.log('After ', action.type);
  return returnVal;
};


const store = createStore(
  combineReducers({films, mode, search}),
  (localStorage['redux-store'])? JSON.parse(localStorage['redux-store']) : stateData,
  applyMiddleware(logger)
);


store.subscribe(() => {
  localStorage['redux-store'] = JSON.stringify(store.getState())
});

//console.log(store.getState());
localStorage['redux-store'] = JSON.stringify(store.getState());

export default store;