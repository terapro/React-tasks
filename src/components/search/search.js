import React, {Component} from 'react';
import {connect} from 'react-redux';
import {startSearch} from 'src/actions';
import PropTypes from 'prop-types';
import {Input} from 'src/components/common/input/input.js';
import {Title} from 'src/components/common/title/title.js';
import {TypeSection} from 'src/components/search/type-section/type-section.js';
import {ButtonSection} from 'src/components/search/button-section/button-section.js';

export class SearchChild extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchTypes: {
        list: ['title', 'genres'],
        active: 'title'
      },
      currentInputValue: ''
    };
    this.startSearching = () => {
      this.props.onSearch(this.state.currentInputValue, this.state.searchTypes.active);
      this.setState({currentInputValue: ''})
    };
    this.enterKeyPressedOnInput = (e) => {
      if (e.key === 'Enter') {
        this.startSearching();
      }
    };
    this.changeSearchType = (newType) => {
      this.setState({
        searchTypes: {
          list: this.state.searchTypes.list, // living as it is
          active: newType // Toggle to the active type
        }
      });
    };
    this.cloneSearchInput = (el) => {
      this.setState({currentInputValue: el.target.value});
    }
  }

  render() {
    const {searchMode} = this.props;
    if (searchMode) {
      return (
        <div className={'header-search'}>
          <Title content='Find your movie' ttlWhite ttlUC/>

          <Input onStartTyping={this.cloneSearchInput}
                 value={this.state.currentInputValue}
                 onKeyEnterPressed={this.enterKeyPressedOnInput}/>

          <div className={'search-components'}>
            <TypeSection searchTypes={this.state.searchTypes}
                         searchTypeClick={this.changeSearchType}/>

            <ButtonSection parentFormId={'search-form'}
                           searchButtonClick={this.startSearching}/>
          </div>
        </div>
      );
    } else {
      return null;
    }
  }
}

SearchChild.propTypes = {
  onSearch: PropTypes.func.isRequired,
  searchMode: PropTypes.bool

};
SearchChild.defaultProps = {
  searchMode: true
};

export const Search = connect(
  store =>
    ({
      searchMode: store.mode.search
    }),
  dispatch =>
    ({
      onSearch(phrase, type) {// Request API
        dispatch(startSearch(phrase, type))
      }
    })
)(SearchChild);