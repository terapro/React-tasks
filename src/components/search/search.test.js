import React from 'react';
import store from "src/store";
import {Provider} from 'react-redux';
import {shallow} from 'enzyme';
import {Search, SearchChild} from 'src/components/search/search.js';

const mockFunc = jest.fn();
const mockSearchPhraseText = 'star wars';
const mockEl = {target: {innerHTML: 'genres', value: 'some input'}, key: 'Enter'};

describe('<Search />', () => {
  describe('Rendering', () => {
    it('should render the component in the search mode', () => {
      const wrapper = shallow(<Provider store={store}><Search onSearch={mockFunc} searchMode/></Provider>);
      expect(wrapper).toMatchSnapshot();
    });
    it('should render the component with default props', () => {
      const wrapper = shallow(<Provider store={store}><Search onSearch={mockFunc} searchMode/></Provider>);
      expect(wrapper).toMatchSnapshot();
    });
    it('should not render the component with disabled searchMode', () => {
            expect(shallow(<Provider store={store}><Search onSearch={mockFunc}/></Provider>).is('.header-search')).toBe(false);
    });
  });
  describe('Methods', () => {
    let instance;
    beforeAll(() => {
          instance = shallow(<SearchChild onSearch={mockFunc}/>).instance();
    });
    afterEach(() => {
      jest.clearAllMocks();
    });
    describe('startSearching', () => {
      it('should reset currentInputValue in the state', () => {
        instance.state = {
          currentInputValue: mockSearchPhraseText,
          searchTypes: {
            list: ['title', 'genres'],
            active: 'title'
          }
        };
        instance.startSearching();
        expect(instance.state.currentInputValue).toEqual('');
      });
    });
    describe('changeSearchType', () => {
      it('should set searchType in the state', () => {
        instance.state = {
          searchTypes: {
            list: ['title', 'genres'],
            active: 'title'
          }
        };
        instance.changeSearchType(mockEl);
        expect(instance.state.searchTypes.active).toEqual(mockEl.target.innerHTML);
      });
    });
    describe('cloneSearchInput', () => {
      it('should set currentInputValue in the state', () => {
        instance.state = {
          currentInputValue: ''
        };
        instance.cloneSearchInput(mockEl);
        expect(instance.state.currentInputValue).toEqual(mockEl.target.value);
      });
    });
    describe('enterKeyPressedOnInput', () => {
      it('should reset currentInputValue in the state', () => {
        instance.state = {
          currentInputValue: mockSearchPhraseText,
          searchTypes: {
            list: ['title', 'genres'],
            active: 'title'
          }
        };
        instance.enterKeyPressedOnInput(mockEl);
        expect(instance.state.currentInputValue).toEqual('');
      });
    });
  });
});
