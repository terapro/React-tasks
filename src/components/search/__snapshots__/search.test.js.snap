// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> Rendering should render the component 1`] = `
<div
  className="header-search"
  id="search-form"
>
  <Title
    content="Find your movie"
    ttlBig={false}
    ttlRed={false}
    ttlUC={true}
    ttlWhite={true}
  />
  <Input
    onKeyPressedFunc={[Function]}
    placeholder=""
    searchInputCallback={[Function]}
    value=""
  />
  <div
    className="search-components"
  >
    <TypeSection
      name="searchby"
      searchTypeCallback={[Function]}
      searchTypes={
        Object {
          "active": "title",
          "list": Array [
            "title",
            "genres",
          ],
        }
      }
    />
    <ButtonSection
      parentFormId="search-form"
      searchButtonCallBack={[Function]}
    />
  </div>
</div>
`;
